Only `SyncClosedLogs` for multiple CFs

solicomo opened the pull request:
Call `SyncClosedLogs()` only if there are more than one column families.

Update several unit tests (in `fault_injection_test` and `db_flush_test`) correspondingly.

See #3840 for more info.

riversand963 commented:
Nice. Also cc @ajkr

riversand963 commented:
#4416 is also duplicate of this. I will close #4416 after merging this one.

solicomo commented:
> #4416 is also duplicate of this. I will close #4416 after merging this one.

Sorry. I didn't know you are working on it.

riversand963 commented:
I took a pass, and found that you removed the case of single column family, which should fail according to the test result of #4416 . Can you add a test for single column family case, too?

solicomo commented:
> I took a pass, and found that you removed the case of single column family, which should fail according to the test result of #4416 . Can you add a test for single column family case, too?

Are you saying test `DBFlushTest::SyncFail`?

riversand963 commented:
> > I took a pass, and found that you removed the case of single column family, which should fail according to the test result of #4416 . Can you add a test for single column family case, too?
> 
> Are you saying test `DBFlushTest::SyncFail`?

Yes. Original implementation always calls `SyncClosedLogs` since the number of column families is always greater than 0. This PR changed the logic such that `SyncClosedLogs` is not called when there is only one column family. Therefore, we are interested in what would happen if there is only one column family in addition to your updated version of the test.

solicomo commented:
> > > I took a pass, and found that you removed the case of single column family, which should fail according to the test result of #4416 . Can you add a test for single column family case, too?
> > 
> > 
> > Are you saying test `DBFlushTest::SyncFail`?
> 
> Yes. Original implementation always calls `SyncClosedLogs` since the number of column families is always greater than 0. This PR changed the logic such that `SyncClosedLogs` is not called when there is only one column family. Therefore, we are interested in what would happen if there is only one column family in addition to your updated version of the test.

Agreed. I'm working on it.
Thanks for your suggestion.

solicomo commented:
Hi @riversand963 , the unit test is added. Could you please review it again?
Thanks.

riversand963 commented:
Hi @solicomo , thanks for working on this. I noticed that you added a column family 'pikachu' in `DBFlushTest.SyncFail`. Therefore, in `DBFlushTest.SyncFail`, the number of column families now is 2 ('default' and 'pikachu'), and `SyncClosedLogs` is still always called.
Also, is it necessary to add one column family in `FaultInjectionTest`? The change you made in `db_impl_compaction_flush.cc`  will mostly affect cases in which there is only one column family.

solicomo commented:
> Hi @solicomo , thanks for working on this. I noticed that you added a column family 'pikachu' in `DBFlushTest.SyncFail`. Therefore, in `DBFlushTest.SyncFail`, the number of column families now is 2 ('default' and 'pikachu'), and `SyncClosedLogs` is still always called.
> Also, is it necessary to add one column family in `FaultInjectionTest`? The change you made in `db_impl_compaction_flush.cc` will mostly affect cases in which there is only one column family.

I think `DBFlushTest.SyncFail` is trying to test the behavior of RocksDB when the filesystem is inactive.
That's why `fault_injection_env` is used and set to be inactive before Sync.

From the `SyncPoint`s in `DBFlushTest.SyncFail`, we can also know that `SyncClosedLogs()` is expected to be called.

To make sure `SyncClosedLogs()` can be called, we need at least two column families.

Similar thing to the tests in `FaultInjectionTest`.

I hope I understand the code right. Please let me know if I am wrong.
Thank you.

riversand963 commented:
@solicomo Due to the sync points dependency in `DBFlushTest.SyncFail`, I agree that adding a column family here is necessary to make it pass, otherwise the test will wait forever.
However, in `FaultInjectionTest`, can you try not add extra column family? My local run (also #4416) shows that, with the proposed change to `FlushMemTableToOutputFile`, `FaultInjectionTest` fails because certain logs are missing. They are missing because we no longer sync closed logs with the new change since there is only one column family. This means `FaultInjectionTest` catches a bug, which is a good thing. The reason why your PR passes this test is because you have two column families, thus RocksDB still sync the closed logs. I don't think we can change this test so that only one code path is followed, while the other causes problem. Please let me know if I have missed anything. Thanks

solicomo commented:
> @solicomo Due to the sync points dependency in `DBFlushTest.SyncFail`, I agree that adding a column family here is necessary to make it pass, otherwise the test will wait forever.
> However, in `FaultInjectionTest`, can you try not add extra column family? My local run (also #4416) shows that, with the proposed change to `FlushMemTableToOutputFile`, `FaultInjectionTest` fails because certain logs are missing. They are missing because we no longer sync closed logs with the new change since there is only one column family. This means `FaultInjectionTest` catches a bug, which is a good thing. The reason why your PR passes this test is because you have two column families, thus RocksDB still sync the closed logs. I don't think we can change this test so that only one code path is followed, while the other causes problem. Please let me know if I have missed anything. Thanks

Are you saying, even `SyncClosedLogs()` is not called, the "missing logs" are still expected to exist?
If they don't exist, there must be a bug and we need to figure out why they don't exist.
Do I understand right?

riversand963 commented:
Sorry for the confusion.
I suggested that we do not add an extra column family to force the control flow of the program to take a specific branch.
If you try this suggestion, you can see that the tests will fail because certain files are missing. We need to understand whether these missing files will hurt the reliability of the database.

solicomo commented:
> Sorry for the confusion.
> I suggested that we do not add an extra column family to force the control flow of the program to take a specific branch.
> If you try this suggestion, you can see that the tests will fail because certain files are missing. We need to understand whether these missing files will hurt the reliability of the database.

OK, got it.
I'm gonna investigate why we lost those log files.
Thank you.

solicomo commented:
Status Update: I am still working on it. It's almost there.

I'm sorry for the delay. But don't blame me, blame Zelda and Nintendo.

solicomo commented:
I rebased and squashed my commits and reworded the commit message.

@riversand963 Could you take a look again?

The reason of test failure `DBFlushTest.SyncFail` is `FaultInjectionTestEnv` fails to untrack the files already deleted.
See summary in commit message for details.

@siying Could you please take a look, too? Because it is related to `DBWALTestWithEnrichedEnv.SkipDeletedWALs`.

Thanks.

siying commented on db/db_wal_test.cc:
```diff
@@ -88,6 +88,7 @@ class DBWALTestWithEnrichedEnv : public DBTestBase {
     enriched_env_ = new EnrichedSpecialEnv(env_->target());
     auto options = CurrentOptions();
     options.env = enriched_env_;
+    options.allow_2pc = true;
```
Why is it added?

solicomo replied:
[On line 535 in `dp_impl_open.cc`](https://github.com/facebook/rocksdb/blob/d59549298fcd50e4aaa59af3dfc039d9a4db5623/db/db_impl_open.cc#L535) we skip WAL files based on `versions_->min_log_number_to_keep_2pc()`.

If I am correct, `min_log_number_to_keep_2pc_` is set only in 2PC mode.

See comment [on line 2976 in `version_set.cc`](https://github.com/facebook/rocksdb/blob/d59549298fcd50e4aaa59af3dfc039d9a4db5623/db/version_set.cc#L2976) .


[On line 3926 in `version_set.cc`](https://github.com/facebook/rocksdb/blob/d59549298fcd50e4aaa59af3dfc039d9a4db5623/db/version_set.cc#L), it is set as `version_edit::min_log_number_to_keep_`, which is set only if `options.allow_2pc` is true.

See [line 389 in `memtable_list.cc`](https://github.com/facebook/rocksdb/blob/d59549298fcd50e4aaa59af3dfc039d9a4db5623/db/memtable_list.cc#L389).

So if we want to "SkipDeletedWALs", we have to enable `options.allow_2pc`.

Please correct me if I am wrong.

Thank you.

siying replied:
But how does it related to your change?

solicomo replied:
`SkipDeletedWALs` should fail because `options.allow_2pc` is not enabled.
It passed before only because there is a extra '/' in the name of WAL file when it is deleted.

For example, when `EnrichedSpecialEnv::DeleteFile()` is called, the file name is `testrocksdb-26040/xxx_2273688893793634580//000003.log`.
When `EnrichedSpecialEnv::NewSequentialFile()` is called, the file name is `testrocksdb-26040/xxx_2273688893793634580/000003.log `.

The same file has two different names, the first one has an extra '/'.
That's why [line 31 in `db_wal_test.cc `](https://github.com/facebook/rocksdb/blob/d59549298fcd50e4aaa59af3dfc039d9a4db5623/db/db_wal_test.cc#L31) is not hit.

`SkipDeletedWALs ` passed "by mistake".

After I fixed the extra '/' issue for `FaultInjectionTestSplitted.FaultTest`, this test failed.
Then I found we have to enable `options.allow_2pc` for this test.

siying replied:
Let me take a closer look. We should handle log gap correctly no matter 2pc is on or off.

siying replied:
You are right. This is not supposed to pass. Thank you for fixing it.

siying commented on db/db_impl_files.cc:
```diff
@@ -456,7 +456,9 @@ void DBImpl::PurgeObsoleteFiles(JobContext& state, bool schedule_only) {
-      fname = dir_to_sync + "/" + to_delete;
+      fname = dir_to_sync
+            + (dir_to_sync.back() == '/' || to_delete.front() == '/' ? "" : "/")
+            + to_delete;
```
assert dir_to_sync and to_delete are not empty.
Actually, maybe dir_to_sync can be empty?

siying replied:
This is my last comment. If you address this, I'll defer to @riversand963 to accept and merge.

solicomo replied:
A new commit is pushed.

Actually, I think maybe we need something like `canonicalize_file_name()` or `realpath()`.

Thanks for your review.

riversand963 commented:
Good job! @solicomo, thanks for the improvement you make. Let me import your change for our internal build/check. Once that's done, we can merge it.
